File: backend/src/main/java/com/suplr/backend/controller/BroadcastController.java
package com.suplr.backend.controller;

import com.suplr.backend.config.Constants;
import com.suplr.backend.dto.BroadcastDtos.BroadcastRequest;
import com.suplr.backend.dto.BroadcastDtos.BroadcastResponse;
import com.suplr.backend.entity.Client;
import com.suplr.backend.entity.Supplier;
import com.suplr.backend.repository.ClientRepository;
import com.suplr.backend.repository.WhatsAppConnectionRepository;
import com.suplr.backend.service.StorageService;
import com.suplr.backend.service.WhatsAppSenderService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

@Slf4j
@RestController
@RequestMapping("/broadcast")
@RequiredArgsConstructor
public class BroadcastController {

    private final WhatsAppConnectionRepository connectionRepository;
    private final ClientRepository clientRepository;
    private final WhatsAppSenderService senderService;
    private final StorageService storageService;

    @PostMapping
    public BroadcastResponse sendBroadcast(
            @AuthenticationPrincipal Supplier supplier,
            @Valid @RequestBody BroadcastRequest req
    ) {
        var conn = connectionRepository.findBySupplierId(supplier.getId()).orElse(null);
        if (conn == null) return BroadcastResponse.empty();

        var clients = clientRepository.findBySupplierId(supplier.getId());
        if (clients.isEmpty()) return BroadcastResponse.empty();

        List<String> numbers = clients.stream()
                .map(Client::getWhatsappNumber)
                .toList();

        if (req.scheduledAt() != null) {
            String jobId = senderService.enqueueBroadcast(
                    conn.getBspEndpoint(), conn.getBspApiKey(),
                    numbers, req.message(), req.scheduledAt(), req.mediaUrl());

            log.info("Broadcast scheduled for {} — job {} (supplier {})",
                    req.scheduledAt(), jobId, supplier.getId());

            return new BroadcastResponse(0, 0, clients.size(), true, jobId);
        }

        List<CompletableFuture<Boolean>> futures = numbers.stream()
                .map(number -> CompletableFuture.supplyAsync(() -> {
                    try {
                        senderService.sendMessage(
                                conn.getBspEndpoint(), conn.getBspApiKey(),
                                number, req.message(), req.mediaUrl());
                        return true;
                    } catch (Exception e) {
                        log.error("Broadcast failed for {}: {}", number, e.getMessage());
                        return false;
                    }
                }, Constants.EXECUTOR))
                .toList();

        List<Boolean> results = futures.stream()
                .map(CompletableFuture::join)
                .toList();

        int sent = (int) results.stream().filter(r -> r).count();
        int failed = results.size() - sent;

        log.info("Broadcast: {}/{} sent (supplier {})", sent, clients.size(), supplier.getId());
        return new BroadcastResponse(sent, failed, clients.size(), false, null);
    }

    @PostMapping("/upload")
    public java.util.Map<String, String> uploadMedia(
            @AuthenticationPrincipal Supplier supplier,
            @RequestParam("file") MultipartFile file
    ) throws IOException {
        byte[] data = file.getBytes();
        String ext = getExtension(file.getOriginalFilename());
        String key = "broadcasts/" + supplier.getId() + "/"
                + UUID.randomUUID().toString().replace("-", "") + ext;
        String url = storageService.upload(
                key, data,
                file.getContentType() != null ? file.getContentType() : "application/octet-stream");
        return java.util.Map.of("url", url);
    }

    private static String getExtension(String filename) {
        if (filename == null || filename.isBlank()) return ".bin";
        int dot = filename.lastIndexOf('.');
        return dot >= 0 ? filename.substring(dot).toLowerCase() : ".bin";
    }
}

File: backend/src/broadcast/broadcast.controller.ts
import {
  Controller, Post, UseGuards, Body, UseInterceptors, UploadedFile,
  Logger,
} from '@nestjs/common';
import { FileInterceptor } from '@nestjs/platform-express';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { IsOptional, IsString, IsDateString } from 'class-validator';
import { JwtAuthGuard } from '../common/guards/jwt-auth.guard.js';
import { CurrentSupplier } from '../common/decorators/current-supplier.decorator.js';
import { Supplier } from '../entities/supplier.entity.js';
import { Client } from '../entities/client.entity.js';
import { WhatsAppConnection } from '../entities/whatsapp-connection.entity.js';
import { WhatsAppSenderService } from '../whatsapp/whatsapp-sender.service.js';
import { StorageService } from '../storage/storage.service.js';
import { randomBytes } from 'node:crypto';

class BroadcastRequest {
  @IsString() message!: string;
  @IsOptional() @IsDateString() scheduledAt?: string;
  @IsOptional() @IsString() mediaUrl?: string;
}

@UseGuards(JwtAuthGuard)
@Controller('broadcast')
export class BroadcastController {
  private readonly logger = new Logger(BroadcastController.name);

  constructor(
    @InjectRepository(WhatsAppConnection) private readonly connRepo: Repository<WhatsAppConnection>,
    @InjectRepository(Client) private readonly clientRepo: Repository<Client>,
    private readonly sender: WhatsAppSenderService,
    private readonly storage: StorageService,
  ) {}

  @Post()
  async sendBroadcast(@CurrentSupplier() supplier: Supplier, @Body() req: BroadcastRequest) {
    const conn = await this.connRepo.findOne({ where: { supplierId: supplier.id } });
    if (!conn) return { sent: 0, failed: 0, total: 0, scheduled: false, jobId: null };

    const clients = await this.clientRepo.find({ where: { supplierId: supplier.id } });
    if (clients.length === 0) return { sent: 0, failed: 0, total: 0, scheduled: false, jobId: null };

    const numbers = clients.map(c => c.whatsappNumber);

    if (req.scheduledAt) {
      const jobId = await this.sender.enqueueBroadcast(
        conn.bspEndpoint, conn.bspApiKey,
        numbers, req.message,
        new Date(req.scheduledAt),
        req.mediaUrl,
      );
      this.logger.log(`Broadcast scheduled for ${req.scheduledAt} — job ${jobId} (supplier ${supplier.id})`);
      return { sent: 0, failed: 0, total: clients.length, scheduled: true, jobId };
    }

    const results = await Promise.allSettled(
      numbers.map(number =>
        this.sender.sendMessage(conn.bspEndpoint, conn.bspApiKey, number, req.message, req.mediaUrl)
      )
    );

    const sent = results.filter(r => r.status === 'fulfilled').length;
    const failed = results.length - sent;
    this.logger.log(`Broadcast: ${sent}/${clients.length} sent (supplier ${supplier.id})`);
    return { sent, failed, total: clients.length, scheduled: false, jobId: null };
  }

  @Post('upload')
  @UseInterceptors(FileInterceptor('file'))
  async uploadMedia(@CurrentSupplier() supplier: Supplier, @UploadedFile() file: Express.Multer.File & { buffer: Buffer }) {
    const ext = getExtension(file.originalname);
    const key = `broadcasts/${supplier.id}/${randomBytes(16).toString('hex')}${ext}`;
    const url = await this.storage.upload(key, file.buffer, file.mimetype ?? 'application/octet-stream');
    return { url };
  }
}

function getExtension(filename?: string): string {
  if (!filename) return '.bin';
  const dot = filename.lastIndexOf('.');
  return dot >= 0 ? filename.substring(dot).toLowerCase() : '.bin';
}
